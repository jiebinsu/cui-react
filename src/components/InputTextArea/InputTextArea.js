import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import "./inputTextArea.scss";

const InputTextArea = forwardRef(
  (
    {
      id,
      name,
      cols,
      rows,
      className,
      label,
      hint,
      hasError,
      errorMsg,
      disabled,
      ...props
    },
    ref
  ) => {
    const hintId = `${id}-hint`;
    const errorId = `${id}-error`;
    const classNames = ["coop-form__field", "coop-form__textarea"];
    const ariaDescribedBy = [];

    if (hint) ariaDescribedBy.push(hintId);

    if (hasError) {
      classNames.push("coop-form__invalid");
      ariaDescribedBy.push(errorId);
    }

    classNames.push(className);

    return (
      <div className="coop-form__row">
        {label && (
          <label htmlFor={id} className="coop-form__label">
            {label}
          </label>
        )}
        {hint && (
          <p id={`${id}-hint`} className="coop-form__hint">
            {hint}
          </p>
        )}
        {hasError && (
          <p id={`${id}-error`} className="coop-form__error">
            {errorMsg}
          </p>
        )}
        <textarea
          cols={cols}
          rows={rows}
          name={name}
          id={id}
          className={classNames.join(" ")}
          aria-describedby={ariaDescribedBy && ariaDescribedBy.join(" ")}
          disabled={disabled}
          ref={ref}
          {...props}
        />
      </div>
    );
  }
);

InputTextArea.defaultProps = {
  cols: 30,
  rows: 4,
  className: null,
  label: null,
  hint: null,
  hasError: false,
  errorMsg: null,
  disabled: false,
};

InputTextArea.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  cols: PropTypes.number,
  rows: PropTypes.number,
  className: PropTypes.string,
  label: PropTypes.string,
  hint: PropTypes.string,
  hasError: PropTypes.bool,
  errorMsg: PropTypes.string,
  disabled: PropTypes.bool,
};

export default InputTextArea;
