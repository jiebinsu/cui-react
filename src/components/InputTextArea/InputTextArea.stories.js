import React from "react";
import InputTextArea from "./InputTextArea";

export default {
  title: "Coop UI/InputTextArea",
  component: InputTextArea,
};

export const Basic = (args) => <InputTextArea {...args} />;
Basic.args = { id: "basic", name: "basic" };

export const All = () => (
  <>
    <InputTextArea id="label" name="label" label="Input with label" />
    <InputTextArea id="label" name="label" hint="Input with hint" />
    <InputTextArea
      id="label"
      name="label"
      label="Input with label and hint"
      hint="Input with label and hint"
    />
  </>
);

export const WithError = () => (
  <>
    <InputTextArea
      id="label"
      name="label"
      label="Input with label"
      errorMsg="Error message"
      hasError
    />
    <InputTextArea
      id="label"
      name="label"
      hint="Input with hint"
      errorMsg="Error message"
      hasError
    />
    <InputTextArea
      id="label"
      name="label"
      label="Input with label and hint"
      hint="Input with label and hint"
      errorMsg="Error message"
      hasError
    />
  </>
);
