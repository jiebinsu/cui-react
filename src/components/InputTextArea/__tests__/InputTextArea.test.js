import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputTextArea from "../index";

describe("InputTextArea", () => {
  it("should render correctly", () => {
    const wrapper = mount(<InputTextArea id="test" name="test" label="test" />);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support all variants", () => {
    const Variants = () => (
      <div>
        <InputTextArea id="label" name="label" label="Input with label" />
        <InputTextArea id="label" name="label" hint="Input with hint" />
        <InputTextArea
          id="label"
          name="label"
          label="Input with label and hint"
          hint="Input with label and hint"
        />
      </div>
    );

    const tree = renderer.create(<Variants />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<Variants />);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support all variants with error", () => {
    const VariantsWithError = () => (
      <div>
        <InputTextArea
          id="label"
          name="label"
          label="Input with label"
          errorMsg="Error message"
          hasError
        />
        <InputTextArea
          id="label"
          name="label"
          hint="Input with hint"
          errorMsg="Error message"
          hasError
        />
        <InputTextArea
          id="label"
          name="label"
          label="Input with label and hint"
          hint="Input with label and hint"
          errorMsg="Error message"
          hasError
        />
      </div>
    );

    const tree = renderer.create(<VariantsWithError />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<VariantsWithError />);
    expect(() => wrapper.unmount()).not.toThrow();
  });
});
