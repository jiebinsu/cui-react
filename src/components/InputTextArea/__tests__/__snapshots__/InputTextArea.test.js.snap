// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextArea should support all variants 1`] = `
<div>
  <div
    className="coop-form__row"
  >
    <label
      className="coop-form__label"
      htmlFor="label"
    >
      Input with label
    </label>
    <textarea
      aria-describedby=""
      className="coop-form__field coop-form__textarea "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
  <div
    className="coop-form__row"
  >
    <p
      className="coop-form__hint"
      id="label-hint"
    >
      Input with hint
    </p>
    <textarea
      aria-describedby="label-hint"
      className="coop-form__field coop-form__textarea "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
  <div
    className="coop-form__row"
  >
    <label
      className="coop-form__label"
      htmlFor="label"
    >
      Input with label and hint
    </label>
    <p
      className="coop-form__hint"
      id="label-hint"
    >
      Input with label and hint
    </p>
    <textarea
      aria-describedby="label-hint"
      className="coop-form__field coop-form__textarea "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
</div>
`;

exports[`InputTextArea should support all variants with error 1`] = `
<div>
  <div
    className="coop-form__row"
  >
    <label
      className="coop-form__label"
      htmlFor="label"
    >
      Input with label
    </label>
    <p
      className="coop-form__error"
      id="label-error"
    >
      Error message
    </p>
    <textarea
      aria-describedby="label-error"
      className="coop-form__field coop-form__textarea coop-form__invalid "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
  <div
    className="coop-form__row"
  >
    <p
      className="coop-form__hint"
      id="label-hint"
    >
      Input with hint
    </p>
    <p
      className="coop-form__error"
      id="label-error"
    >
      Error message
    </p>
    <textarea
      aria-describedby="label-hint label-error"
      className="coop-form__field coop-form__textarea coop-form__invalid "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
  <div
    className="coop-form__row"
  >
    <label
      className="coop-form__label"
      htmlFor="label"
    >
      Input with label and hint
    </label>
    <p
      className="coop-form__hint"
      id="label-hint"
    >
      Input with label and hint
    </p>
    <p
      className="coop-form__error"
      id="label-error"
    >
      Error message
    </p>
    <textarea
      aria-describedby="label-hint label-error"
      className="coop-form__field coop-form__textarea coop-form__invalid "
      cols={30}
      disabled={false}
      id="label"
      name="label"
      rows={4}
    />
  </div>
</div>
`;
